### DISCUSSION TITLE
Callback after reconnectionAttempts reached?

### DISCUSSION THREAD

**Discussion by jhogg11** [socketio/socket.io#3976] [Q&A] [answered]
I'm working on a project using Socket.io with React and Express. The site will display data to users that is periodically updated and I'm trying to account for any situation where the back end goes down. Rather than displaying stale data, I'd prefer to show an error message.

I see that there's a `reconnect_failed` event, however, if I test it out by manually killing the server, it doesn't fire. Additionally, `reconnect_attempt` does not fire on each connection attempt (which I can see in the console). It's a little confusing because most `.on` events are received while the socket is connected, however, these two events assume a disconnected socket.

So I assume there's something I'm not understanding about how these events work or how events in general work.

Here's my current code:

```
const socket = io("http://localhost:8080", {
    reconnectionDelay: 5000,
    reconnectionAttempts: 2
});
socket.on("reconnect_attempt", function() {
    console.log("reconnecting")
});
socket.on('reconnect_failed', function() {
    console.log("reconnect failed");
});  
```
I can get around the issue by setting a variable and a timeout on `socket.on('disconnect')`, changing the variable if the client successfully reconnects, and if not, calling the timeout callback to show the error message (the timeout duration would be based on `reconnectionDelay * reconnectionAttempts`).

But that's pretty messy, so I'd like to have a callback that gets called automatically after `reconnectionAttempts` is reached.  Is there a built-in way to do this?

**Comment by darrachequesne** [comment, marked as answer]
Which version of `socket.io-client` are you using?

Starting with v3, the socket instance no longer emits events related to reconnection attempts, you must use:

```js
socket.io.on("reconnect_attempt", function() {
    console.log("reconnecting")
});

socket.io.on('reconnect_failed', function() {
    console.log("reconnect failed");
});
```

Does that explain your issue?

Reference:

- https://socket.io/docs/v4/client-socket-instance/#Events
- https://socket.io/docs/v4/migrating-from-2-x-to-3-0/#The-Socket-instance-will-no-longer-forward-the-events-emitted-by-its-Manager

**Reply by jhogg11** [reply]
I'm using version 4.1 and that does the trick.  Thanks for your help.
